cocotb/triggers.py

```diff
@@ -729,3 +729,42 @@ def _wait(self):
         for _ in range(self.num_cycles):
             yield trigger
         raise ReturnValue(self)
+
+
+@decorators.coroutine
+def with_timeout(triggers, timeout_time, timeout_unit=None):
+    """
+    Waits on triggers, throws an exception if it waits longer than the given time
+
+    Can be used with any expression that would otherwise appear to the right of a
+    `yield` (or `await` in Python 3) statement.
+
+    Usage:
+
+    .. code-block:: python
+
+        yield with_timeout(coro, 100, 'ns')
+        yield with_timeout([coro, event.wait(), ...], 100, 'ns')
+
+    Args:
+        triggers (cocotb_yield_expr):
+            Any expression that could be right of a yield statement in cocotb
+        timeout_time (int):
+            Time duration
+        timeout_unit (str, optional):
+            Units of duration, accepts any values that :class:`cocotb.triggers.Timer` does
```

Comment:
Could use a `~` here to make this render as `Timer` rather than `cocotb.triggers.Timer`, while still linking correctly:

```suggestion
            Units of duration, accepts any values that :class:`~cocotb.triggers.Timer` does
```